### PR TITLE
Add Alibaba Cloud ECI provider

### PR DESCRIPTION
**Pull request by jovizhangwei**
Alibaba Cloud ECI(Elastic Container Instance) is a service that allow you
run containers without having to manage servers or clusters.

This commit add ECI provider for virtual kubelet, connects ECI with
kubernetes cluster.

Signed-off-by: xianwei.zw <xianwei.zw@alibaba-inc.com>
Signed-off-by: shidao.ytt <shidao.ytt@alibaba-inc.com>

### CODE REVIEW

**Comment by resouer** [comment]
@jovizhangwei Would you be able to split the PR into several commits please?

For example:

1. code
2. docs & examples
3. vendor

It would make review easier and win a quick merge :-)

**Comment by jovizhangwei** [comment]
Hi @resouer,

Thanks for reviewing, actually the layout of this commit is quite simple :). Maybe it's more clean to put in one commit?

1. core code: providers/alicloud/eci.go
2. readme: providers/alicloud/README.md
3. vendor directory: vendor/github.com/aliyun/alibaba-cloud-sdk-go

**Comment by resouer** [comment]
It's not mandatory, while Github is not friendly for big PRs, I think you may noticed some source code files have been folded up ...

Also, I noticed there are some Chinese characters in eci.svg, I am not sure why this is needed but you may want to do a clean up :-)

**Comment by jovizhangwei** [comment]
Sure, we will fix it asap. Thanks.

File: providers/alicloud/eci.go [hunk 1]
```diff
@@ -0,0 +1,852 @@
...
+	request.Tags = tags
+	request.SecurityGroupId = p.secureGroup
+	request.VSwitchId = p.vSwitch
+	request.ContainerGroupName = ContainerGroupName
+	response, err := p.eciClient.CreateContainerGroup(request)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("CreateContainerGroup successed. %s, %s, %s\n", response.RequestId, response.ContainerGroupId, ContainerGroupName)
```
Comment (robbiezhang) [inline providers/alicloud/eci.go]: switch to use vk log package and add ctx in the log.

**Comment by robbiezhang** [comment]
@shidao-ytt, please sign the linuxfoudation CLA.

Please also update the contributing file to add the owner.
https://github.com/virtual-kubelet/virtual-kubelet/blob/master/CONTRIBUTING.md

File: providers/alicloud/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+## Deployment of the ECI provider in your cluster
+configure and launch virtual kubelet
+```
+export ECI_REGION=cn-hangzhou
+export ECI_SECURITY_GROUP=sg-123
+export ECI_VSWITCH=vsw-123
+export ECI_ACCESS_KEY=123
+export ECI_SECRET_KEY=123
+
+virtual-kubelet --provider alicloud --taint alibabacloud.com/eci
```
[the remark this answers is not in the document]
Comment (shidao-ytt) [reply providers/alicloud/README.md]: updated

File: providers/alicloud/README.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+kind: Pod
+metadata:
+  name: mypod
+spec:
+  nodeName: virtual-kubelet
+  containers:
+  - name: nginx
+    image: nginx
+  tolerations:
+  - key: alibabacloud.com/eci
```
[the remark this answers is not in the document]
Comment (shidao-ytt) [reply providers/alicloud/README.md]: updated.

**Comment by shidao-ytt** [comment]
@robbiezhang all change request is done. the ci/circleci failed is due to baseline

File: providers/alicloud/eci.go [hunk 4]
```diff
@@ -0,0 +1,855 @@
...
+func (p *ECIProvider) DeletePod(ctx context.Context, pod *v1.Pod) error {
+	eciId := ""
+	for _, cg := range p.GetCgs() {
+		if getECITagValue(&cg, "PodName") == pod.Name && getECITagValue(&cg, "NameSpace") == pod.Namespace {
+			eciId = cg.ContainerGroupId
+			break
+		}
+	}
+	if eciId == "" {
+		return errors.New(fmt.Sprintf("DeletePod cann't find Pod %s-%s", pod.Namespace, pod.Name))
```
[the remark this answers is not in the document]
Comment (shidao-ytt) [reply providers/alicloud/eci.go]: the code is updated

File: providers/alicloud/eci.go [hunk 5]
```diff
@@ -0,0 +1,855 @@
...
+	var authConfigs map[string]AuthConfig
+	repoData, ok := secret.Data[string(v1.DockerConfigKey)]
+
+	if !ok {
+		return ips, fmt.Errorf("no dockercfg present in secret")
+	}
+
+	err = json.Unmarshal(repoData, &authConfigs)
+	if err != nil {
+		return ips, err
```
[the remark this answers is not in the document]
Comment (shidao-ytt) [reply providers/alicloud/eci.go]: the code is updated

File: providers/alicloud/eci.go [hunk 6]
```diff
@@ -0,0 +1,855 @@
...
+	return &p, err
+}
+
+// CreatePod accepts a Pod definition and creates
+// an ECI deployment
+func (p *ECIProvider) CreatePod(ctx context.Context, pod *v1.Pod) error {
+	//Ignore daemonSet Pod
+	if pod != nil && pod.OwnerReferences != nil && len(pod.OwnerReferences) != 0 && pod.OwnerReferences[0].Kind == "DaemonSet" {
+		msg := fmt.Sprintf("Skip to create DaemonSet pod %q", pod.Name)
+		log.G(context.TODO()).WithField("Method", "CreatePod").Info(msg)
```
[the remark this answers is not in the document]
Comment (shidao-ytt) [reply providers/alicloud/eci.go]: the code is updated

File: Gopkg.lock [hunk 7]
```diff
@@ -2,13 +2,13 @@
 
 
 [[projects]]
-  digest = "1:ef14b1839bd43a923896a762b31d000c12f0d21506cde7251d9342bb6acb4c05"
+  digest = "1:6453651bd5fc38b3a07d133c295a881c1bce87d4be4ccefe61eee26d22d6b15c"
```
Comment (robbiezhang) [inline Gopkg.lock]: why do you change this?

**Comment by shidao-ytt** [comment]
@cpuguy83 @robbiezhang 
thanks very much for your review comments
all change requests are done.

File: providers/alicloud/eci.go [hunk 8]
```diff
@@ -0,0 +1,855 @@
...
+
+// GetPods returns a list of all pods known to be running within ECI.
+func (p *ECIProvider) GetPods(ctx context.Context) ([]*v1.Pod, error) {
+	pods := make([]*v1.Pod, 0)
+	for _, cg := range p.GetCgs() {
+		c := cg
+		pod, err := containerGroupToPod(&c)
+		if err != nil {
+			msg := fmt.Sprint("error converting container group to pod", cg.ContainerGroupId, err)
+			log.G(context.TODO()).WithField("Method", "GetPods").Info(msg)
```
Comment (cpuguy83) [inline providers/alicloud/eci.go]: context

**Pull request merged**